src/main/scala/io/iohk/ethereum/network/discovery/PeerDiscoveryManager.scala

```diff
@@ -1,201 +1,160 @@
+    discoveryServiceResource: Resource[Task, v4.DiscoveryService]
+)(implicit scheduler: Scheduler)
+    extends Actor
     with ActorLogging {
 
   import PeerDiscoveryManager._
 
-  val expirationTimeSec = discoveryConfig.messageExpiration.toSeconds
-
-  val bootStrapNodesInfo = discoveryConfig.bootstrapNodes.map(DiscoveryNodeInfo.fromNode)
-
-  var pingedNodes: Map[ByteString, PingInfo] = Map.empty
-
-  var nodesInfo: Map[ByteString, DiscoveryNodeInfo] = {
-    val knownNodesURIs =
-      if (discoveryConfig.discoveryEnabled) knownNodesStorage.getKnownNodes()
-      else Set.empty
-    val nodesInfo = knownNodesURIs.map(uri => DiscoveryNodeInfo.fromUri(uri)) ++ bootStrapNodesInfo
-    nodesInfo.map { nodeInfo => nodeInfo.node.id -> nodeInfo }.toMap
-  }
-
-  if (discoveryConfig.discoveryEnabled) {
-    discoveryListener ! DiscoveryListener.Subscribe
-    context.system.scheduler.scheduleWithFixedDelay(
-      discoveryConfig.scanInitialDelay,
-      discoveryConfig.scanInterval,
-      self,
-      Scan
-    )
-  }
-
-  def scan(): Unit = {
-    // Ping a random sample from currently pinged nodes without the answer
-    new Random().shuffle(pingedNodes.values).take(2 * discoveryConfig.scanMaxNodes).foreach { pingInfo =>
-      val node = pingInfo.nodeinfo.node
-      sendPing(Endpoint.makeEndpoint(node.udpSocketAddress, node.tcpPort), node.udpSocketAddress, pingInfo.nodeinfo)
+  // The following logic is for backwards compatibility.
+  val alreadyDiscoveredNodes: Set[Node] =
+    if (!discoveryConfig.reuseKnownNodes) Set.empty
+    else {
+      // The manager considered the bootstrap nodes discovered, even if discovery was disabled.
+      val bootstrapNodes: Set[Node] =
+        discoveryConfig.bootstrapNodes
+      // The known nodes were considered discovered even if they haven't yet responded to pings; unless discovery was disabled.
+      val knownNodes: Set[Node] =
+        if (!discoveryConfig.discoveryEnabled) Set.empty
+        else
+          knownNodesStorage.getKnownNodes().map(Node.fromUri)
+
+      bootstrapNodes ++ knownNodes
     }
 
-    nodesInfo.values.toSeq
-      .sortBy(_.addTimestamp)
-      .takeRight(discoveryConfig.scanMaxNodes)
-      .foreach { nodeInfo =>
-        sendPing(
-          Endpoint.makeEndpoint(nodeInfo.node.udpSocketAddress, nodeInfo.node.tcpPort),
-          nodeInfo.node.udpSocketAddress,
-          nodeInfo
-        )
+  override def receive: Receive = init
+
+  // The service hasn't been started yet, so it just serves the static known nodes.
+  def init: Receive = {
+    case GetDiscoveredNodesInfo =>
+      sendDiscoveredNodesInfo(None, sender)
+
+    case Start =>
+      if (discoveryConfig.discoveryEnabled) {
+        log.info("Starting peer discovery...")
+        discoveryServiceResource.allocated
+          .map { case (discoveryService, release) =>
+            Started(discoveryService, release)
+          }
+          .runToFuture
+          .pipeTo(self)
```

Comment:
If the starting fails an error message is [logged](https://github.com/input-output-hk/mantis/pull/770/files#diff-9f1b31738bca75591cd8d4260e448ab541b27cf3e53a534f3e43ff0f5083ff70R65), but you're right it would be nicer if the manager wasn't left in a transient state, I'll change it so it gets notified about failures.

Reply:
Done. Crazy amount of boilerplate, it's nicer in the other project where everything is a resource.